crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -0,0 +1,143 @@
+use crate::{
+    domain::competition::order,
+    tests::{
+        self,
+        setup::{
+            ab_adjusted_pool,
+            ab_liquidity_quote,
+            ab_order,
+            ab_solution,
+            ExpectedOrder,
+            FeePolicy,
+            Test,
+        },
+    },
+};
+
+async fn protocol_fee_test_case(
+    side: order::Side,
+    fee_policy: FeePolicy,
+    expected_order: ExpectedOrder,
+) {
+    let test_name = format!("Protocol Fee: {side:?} {fee_policy:?}");
+    let quote = ab_liquidity_quote()
+        .sell_amount(50000000000000000000u128.into())
+        .buy_amount(40000000000000000000u128.into());
+    let pool = ab_adjusted_pool(quote);
+    let order = ab_order()
+        .kind(order::Kind::Limit)
+        .sell_amount(50000000000000000000u128.into())
+        .side(side)
+        .solver_fee(Some(10000000000000000000u128.into()))
+        .fee_policy(fee_policy)
+        .executed(40000000000000000000u128.into());
```

Comment:
Can we test partial fills as well?

Reply:
That requires some extra refactoring. Wdyt about implementing it in a separate PR since the current one is already too big?